File: Year-2023/LeetCode-March-Challenge/ConvertSortedListtoBinarySearchTree.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    private ArrayList<Integer> res = new ArrayList<>();
    private void toArray(ListNode head){
        ListNode temp = head;
        while(temp != null){
            res.add(temp.val);
            temp = temp.next;
        }
        return;
    }
    private TreeNode arrBST(ArrayList<Integer> res, int low, int high){
        if(low >= high) return null;
        int mid = low + (high - low)/2;
        TreeNode newTree = new TreeNode(res.get(mid));
        newTree.left = arrBST(res, low, mid);
        newTree.right = arrBST(res, mid + 1, high);
        return newTree;
    }
    public TreeNode sortedListToBST(ListNode head) {
        toArray(head);
        return arrBST(res, 0, res.size());
    }
}

public class ConvertSortedListtoBinarySearchTree {
    public static void main(String[] args) {

    }
}

File: Year-2023/LeetCode-March-Challenge/ConvertSortedListtoBinarySearchTree.js
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

var sortedListToBST = function (head) {
    let res = [];

    const toArray = (head) => {
        let temp = head;
        while (temp) {
            res.push(temp.val);
            temp = temp.next;
        }
        return;
    }

    const arrBST = (res, low, high) => {
        if (low >= high) return null;
        let mid = low + Math.trunc((high - low) / 2);
        let newTree = new ListNode(res[mid]);
        newTree.left = arrBST(res, low, mid);
        newTree.right = arrBST(res, mid + 1, high);
        return newTree;
    }

    toArray(head);
    return arrBST(res, 0, res.length);
};